fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/ventilation_control_module.rs

```diff
@@ -217,26 +230,44 @@ impl ForwardCargoVentilationControlSystem {
         acs_overhead: &impl AirConditioningOverheadShared,
         pressurization_overhead: &impl PressurizationOverheadShared,
     ) {
-        // TODO: Add failures and smoke detection
         self.isolation_valves_open_allowed = acs_overhead.fwd_cargo_isolation_valve_is_on()
             && !pressurization_overhead.ditching_is_on()
-            && !active_channel_has_fault;
-        self.extraction_fan_is_on =
-            self.isolation_valves_open_allowed && !pressurization_overhead.ditching_is_on();
+            && !active_channel_has_fault
+            && !self.fwd_isol_valve_failure.is_active();
+        self.extraction_fan_is_on = self.isolation_valves_open_allowed
+            && !pressurization_overhead.ditching_is_on()
+            && !self.fwd_extract_fan_failure.is_active();
     }
 
     fn fwd_extraction_fan_is_on(&self) -> bool {
         self.extraction_fan_is_on
     }
+
     fn fwd_isolation_valves_open_allowed(&self) -> bool {
         self.isolation_valves_open_allowed
     }
+
+    fn fwd_isolation_valve_has_failed(&self) -> bool {
+        self.fwd_isol_valve_failure.is_active()
+    }
+}
+
+impl SimulationElement for ForwardCargoVentilationControlSystem {
+    fn accept<T: SimulationElementVisitor>(&mut self, visitor: &mut T) {
+        self.fwd_isol_valve_failure.accept(visitor);
+        self.fwd_extract_fan_failure.accept(visitor);
+        visitor.visit(self);
+    }
 }
 
 struct BulkVentilationControlSystem {
     duct_heater_on_allowed: bool,
```

Comment:
Would be cool if you could add the power source and power usage for the duct heater too.

Reply:
The cargo heater already had the power source set up in `air_conditioning/mod`. One could argue that the failure could also live within the `AirHeater` struct (instead of the ventillation controller), but I guess it doesn't matter if the item fails or the controller fails, the result is the same!